### PR TITLE
Improve error messages

### PR DESCRIPTION
**Pull request by alexshtin**
Remove duplicate information from failure messages and add more context to go error messages.

### CODE REVIEW

File: internal/error.go [hunk 1]
```diff
@@ -406,9 +420,20 @@ func NewContinueAsNewError(ctx Context, wfn interface{}, args ...interface{}) *C
...
+// Error from error interface.
 func (e *ApplicationError) Error() string {
-	return e.message
+	msg := e.message()
+	if e.errType != "" {
+		msg = fmt.Sprintf("%s (type: %s, retryable: %v)", msg, e.errType, !e.nonRetryable)
+	}
+	if e.cause != nil {
+		msg = fmt.Sprintf("%s: %v", msg, e.cause)
+	}
```
Comment (alexshtin) [inline internal/error.go:431]: This will make error message looks like native error message in case if original error was  standart go error (i.e. `errors.New("message")`).

File: internal/error.go [hunk 2]
```diff
@@ -216,16 +217,24 @@ type (
...
 	}
 
 	failureHolder interface {
 		setFailure(*failurepb.Failure)
 		failure() *failurepb.Failure
 	}
+
+	messenger interface {
+		message() string
+	}
```
Comment (wxing1292) [inline internal/error.go:231]: this is the breaking change we talked about few days ago?

File: internal/error.go [hunk 3]
```diff
@@ -241,9 +250,9 @@ var (
 )
 
 // NewApplicationError create new instance of *ApplicationError with message, type, and optional details.
-func NewApplicationError(message, errType string, nonRetryable bool, cause error, details ...interface{}) *ApplicationError {
+func NewApplicationError(msg, errType string, nonRetryable bool, cause error, details ...interface{}) *ApplicationError {
```
Comment (wxing1292) [inline internal/error.go:253]: can we not reuse the var type? 
like `msg, errType string` vs `msg string, errType string`
former is hard to parse?

File: internal/error.go [hunk 4]
```diff
@@ -769,7 +855,8 @@ func convertFailureToError(failure *failurepb.Failure, dc converter.DataConverte
 	} else if failure.GetTimeoutFailureInfo() != nil {
 		timeoutFailureInfo := failure.GetTimeoutFailureInfo()
 		lastHeartbeatDetails := newEncodedValues(timeoutFailureInfo.GetLastHeartbeatDetails(), dc)
-		err = NewTimeoutError(
+		err = newTimeoutError(
+			failure.GetMessage(),
```
Comment (alexshtin) [inline internal/error.go:859]: `TimeoutFailure` may have message set on server (`WorkflowTask timeout`/`ActivityTask timeout`). This will propagate it to error.

File: internal/error.go (hunk 2 above)
Comment (alexshtin) [reply internal/error.go:231]: It was today morning :-) And actually not. This is just internal interface to simplify dealing with error messages.

**Review by wxing1292 — APPROVED** [review]

File: internal/internal_event_handlers.go [hunk 5]
```diff
@@ -1228,10 +1227,22 @@ func (weh *workflowExecutionEventHandlerImpl) handleStartChildWorkflowExecutionF
...
+			0,
+			enumspb.RETRY_STATE_NON_RETRYABLE_FAILURE,
+			errors.New("workflow execution already started"),
+		)
+		childWorkflow.handle(nil, err)
+		return nil
+	}
 
-	return nil
+	return fmt.Errorf("unknown cause: %v", attributes.GetCause())
```
Comment (alexshtin) [inline internal/internal_event_handlers.go:1245]: I am not sure about this change.

File: internal/error.go [hunk 6]
```diff
@@ -645,6 +718,10 @@ func getErrType(err error) string {
 	for t = reflect.TypeOf(err); t.Kind() == reflect.Ptr; t = t.Elem() {
 	}
 
+	if t.Name() == goErrType {
+		return ""
+	}
```
Comment (alexshtin) [inline internal/error.go:723]: If it was standard error created with `errors.New` error type will be empty string. I think this is better for `Type` field.

**Review by mfateev — APPROVED** [review]
